src/libimg/Image.cpp

```diff
@@ -275,6 +277,227 @@ void Image::drawImage(const Image& src, int dstX, int dstY) {
+    // Due to compute complexity, all filled circles are rendered and cached with a 
+    // key consisting of the radius and color. So typically just a small image blend.
+    uint64_t key = (uint64_t)(radius) << 32 | color;
+    std::shared_ptr<img::Image> pImage;
+    auto it = circleCache.find(key);
+    if (it != circleCache.end()) {
+        pImage = it->second;
+    } else {
+        pImage = std::make_shared<img::Image>(radius*2+1, radius*2+1, 0x00FFFFFF);
+        pImage->fillCircleCacheImage(radius, radius, radius, color); 
+        circleCache.insert(std::make_pair(key, pImage));
+    }
+    blendImage0(*pImage, x_centre-radius, y_centre-radius);
+}
+
+// This class is a representation of a classic line equation ax + by + c
+// It is used in rectangle fills to determine whether a point is inside or outside
+// a rectangle bounding line. And how far away from the line it is so as to determine
+// appropriate alpha blend for anti-aliasing 
+class LineEquation {
+public:
+    LineEquation(int x1, int y1, int x2, int y2, int ref_x, int ref_y) {
+        a = y1 - y2;
+        b = x2 - x1;
+        c = x1*y2 - x2*y1;
+        normalAdjust = 10/sqrt(a*a + b*b);
+        refOffsetSign = (computeDistance(ref_x, ref_y) >= 0);
+    }
+    ~LineEquation(){
+    }
+    int getIntensity(int x, int y) {
+        // Return a value between 0 and 4, which is the multiplicative
+        // contribution to alphas blend from this line instance.
+        int intensity;
+        int distance = (int)(computeDistance(x,y) * normalAdjust);
+        bool sameSign = ((distance >= 0) == refOffsetSign);
+        bool throughPixel = abs(distance) < 7;
+        if (throughPixel) {
+            int offset = abs(distance);
+            if (sameSign)
+                intensity = 2 + (offset*2)/7;
+            else
+                intensity = 2 - (offset*2)/7;
+            return intensity;
+        } else { 
+            if (sameSign)
+                return 4;
+            else
+                return 0;
+        }
+    }
+private:
+    int a;
+    int b;
+    int c;
+    int refOffsetSign;
+    float normalAdjust;
+
+    int computeDistance(int x, int y) {
+        return a * x + b * y + c;
+    }    
+};
+
+// Fill rotated rectangle, given 4 points
+// Points must be in an order where successive points create each one of the bounding lines
+void Image::fillRectangle(int x0, int y0, int x1, int y1, int x2, int y2, int x3, int y3, uint32_t color) {
+    int xCentre = (x0 + x1 + x2 + x3)/4;
+    int yCentre = (y0 + y1 + y2 + y3)/4;
+
+    int xMin = std::min({x0, x1, x2, x3});
+    xMin = std::max(0, xMin);
+    int xMax = std::max({x0, x1, x2, x3});
+    xMax = std::min(width, xMax);
+    int yMin = std::min({y0, y1, y2, y3});
+    yMin = std::max(0, yMin);
+    int yMax = std::max({y0, y1, y2, y3});
+    yMax = std::min(width, yMax);
+
+    std::vector<std::shared_ptr<LineEquation>> lines;

```

Comment:
Drawing lots of rectangles in a loop will result in a lot of heap operations here. Since the line vector is not needed outside of this function and contains a fixed number of lines, I would simply use an array on the stack.